File: new claimsat/claimsat-reunify/src/services/claimScoring.ts
import type { Claim, ClaimScore, Evidence, Disaster } from '../models/types';
import { calculateLocationScore } from '../utils/geo';
import { calculateTimeScore } from '../utils/time';
import {
  calculateEvidenceTypeScore,
  calculateVisualRelevanceScore,
  calculateMetadataIntegrityScore,
} from './evidenceAnalysis';
/**
 * Score weights (must sum to 100)
 */
const WEIGHTS = {
  locationMatch: 30,
  timeProximity: 20,
  evidenceType: 15,
  visualRelevance: 20,
  metadataIntegrity: 15,
};
/**
 * Calculate comprehensive claim score
 * This is the CORE scoring engine - must be stable and explainable
 */
export const calculateClaimScore = (
  claim: Claim,
  disaster: Disaster,
  evidence: Evidence[]
): ClaimScore => {
  // 1. Location Match Score
  const locationResult = calculateLocationScore(claim.location, disaster);
  const locationScore = locationResult.score;
  
  // 2. Time Proximity Score
  const timeResult = calculateTimeScore(
    claim.incidentDate,
    disaster.startDate,
    disaster.endDate
  );
  const timeScore = timeResult.score;
  
  // 3. Evidence Type Score
  const evidenceTypeScore = calculateEvidenceTypeScore(evidence);
  
  // 4. Visual Relevance Score
  const visualRelevanceScore = calculateVisualRelevanceScore(evidence);
  
  // 5. Metadata Integrity Score
  const metadataIntegrityScore = calculateMetadataIntegrityScore(evidence);
  
  // Calculate weighted overall score
  const overall =
    (locationScore * WEIGHTS.locationMatch) / 100 +
    (timeScore * WEIGHTS.timeProximity) / 100 +
    (evidenceTypeScore * WEIGHTS.evidenceType) / 100 +
    (visualRelevanceScore * WEIGHTS.visualRelevance) / 100 +
    (metadataIntegrityScore * WEIGHTS.metadataIntegrity) / 100;
  
  // Clamp to 0-100
  const finalScore = Math.max(0, Math.min(100, overall));
  
  // Generate explanation
  const explanation = generateScoreExplanation(
    finalScore,
    locationResult.explanation,
    timeResult.explanation,
    evidenceTypeScore,
    visualRelevanceScore,
    evidence.length
  );
  
  return {
    overall: Math.round(finalScore * 10) / 10, // Round to 1 decimal
    breakdown: {
      locationMatch: Math.round(locationScore * 10) / 10,
      timeProximity: Math.round(timeScore * 10) / 10,
      evidenceType: Math.round(evidenceTypeScore * 10) / 10,
      visualRelevance: Math.round(visualRelevanceScore * 10) / 10,
      metadataIntegrity: Math.round(metadataIntegrityScore * 10) / 10,
    },
    explanation,
    calculatedAt: new Date().toISOString(),
  };
};
/**
 * Generate human-readable explanation
 */
const generateScoreExplanation = (
  overall: number,
  locationExplanation: string,
  timeExplanation: string,
  evidenceTypeScore: number,
  visualRelevanceScore: number,
  evidenceCount: number
): string => {
  let explanation = '';
  
  // Overall assessment
  if (overall >= 75) {
    explanation += '✓ HIGH CONFIDENCE: This claim shows strong indicators of validity. ';
  } else if (overall >= 50) {
    explanation += '⚠ NEEDS REVIEW: This claim has moderate confidence and requires manual verification. ';
  } else if (overall >= 25) {
    explanation += '⚠ LOW CONFIDENCE: This claim has weak indicators and needs careful review. ';
  } else {
    explanation += '✗ VERY LOW CONFIDENCE: This claim has significant concerns. ';
  }
  
  explanation += '\n\n';
  
  // Location
  explanation += `📍 Location: ${locationExplanation}\n`;
  
  // Time
  explanation += `⏰ Timing: ${timeExplanation}\n`;
  
  // Evidence
  explanation += `📎 Evidence: ${evidenceCount} file(s) submitted`;
  if (evidenceTypeScore >= 75) {
    explanation += ' (strong evidence collection)';
  } else if (evidenceTypeScore >= 50) {
    explanation += ' (moderate evidence)';
  } else {
    explanation += ' (limited evidence)';
  }
  explanation += '\n';
  
  // Visual Analysis
  if (visualRelevanceScore >= 70) {
    explanation += '🔍 Visual Analysis: Evidence appears relevant to disaster context\n';
  } else if (visualRelevanceScore >= 40) {
    explanation += '🔍 Visual Analysis: Evidence relevance is uncertain\n';
  } else {
    explanation += '🔍 Visual Analysis: Evidence may not be disaster-related\n';
  }
  
  explanation += '\n';
  explanation += '⚖️ Final Decision: This is a preliminary assessment. Human authority must review and make final judgment.';
  
  return explanation;
};
/**
 * Derive claim status from score
 */
export const deriveClaimStatus = (
  score: number
): 'pending' | 'approved' | 'needs_review' | 'rejected' => {
  if (score >= 75) return 'approved';
  if (score >= 50) return 'needs_review';
  if (score >= 25) return 'pending';
  return 'rejected';
};

File: new claimsat/backend/services/claim_scoring.py
"""
Claim Scoring Service
Multi-factor scoring engine with NO hard gates
"""
from typing import Dict, Any, Tuple
from models.claim import ClaimScore, ScoringFactors, ClaimStatus, Evidence, EvidenceType
from services.disaster_verification import verify_claim_against_disaster
from services.evidence_analysis import analyze_evidence
from core.config import settings
import logging
logger = logging.getLogger(__name__)
def calculate_evidence_type_score(evidence_list: list) -> Tuple[float, str]:
    """
    Calculate evidence type score
    Videos > Images > No evidence
    """
    if not evidence_list:
        return 0.0, "No evidence provided"
    
    has_video = any(e.get("type") == "video" for e in evidence_list)
    has_image = any(e.get("type") == "image" for e in evidence_list)
    
    if has_video:
        count = sum(1 for e in evidence_list if e.get("type") == "video")
        return 100.0, f"Video evidence provided ({count} video(s))"
    elif has_image:
        count = sum(1 for e in evidence_list if e.get("type") == "image")
        return 75.0, f"Image evidence provided ({count} image(s))"
    else:
        return 50.0, "Evidence provided but type unclear"
def calculate_metadata_integrity_score(evidence_list: list) -> Tuple[float, str]:
    """
    Calculate metadata integrity score
    Checks if evidence has proper metadata (capture time, location)
    """
    if not evidence_list:
        return 0.0, "No evidence to verify metadata"
    
    total_evidence = len(evidence_list)
    has_capture_time = sum(1 for e in evidence_list if e.get("capture_time"))
    has_location = sum(1 for e in evidence_list if e.get("location"))
    
    # Calculate percentage of evidence with metadata
    time_percentage = (has_capture_time / total_evidence) * 100
    location_percentage = (has_location / total_evidence) * 100
    
    avg_score = (time_percentage + location_percentage) / 2
    
    explanation = f"{has_capture_time}/{total_evidence} with timestamps, {has_location}/{total_evidence} with location data"
    
    return avg_score, explanation
async def calculate_claim_score(
    claim_location: Dict[str, float],
    incident_time: str,
    disaster_id: str,
    evidence_list: list,
    visual_analyses: Dict[str, Tuple[float, str]]
) -> ClaimScore:
    """
    Calculate comprehensive claim score using weighted factors
    
    NO HARD GATES - all factors contribute to final score
    """
    
    # 1. Verify against disaster (location + time)
    disaster, location_score, location_explanation, time_score, time_explanation = \
        await verify_claim_against_disaster(claim_location, incident_time, disaster_id)
    
    # 2. Evidence type score
    evidence_type_score, evidence_type_explanation = calculate_evidence_type_score(evidence_list)
    
    # 3. Visual relevance score (aggregate from all evidence)
    if visual_analyses:
        visual_scores = [score for score, _ in visual_analyses.values()]
        avg_visual_score = sum(visual_scores) / len(visual_scores) * 100  # Convert to 0-100
        
        # Collect explanations
        visual_explanations = [exp for _, exp in visual_analyses.values()]
        visual_explanation = "; ".join(visual_explanations[:2])  # First 2 to keep concise
    else:
        avg_visual_score = 50.0
        visual_explanation = "No visual analysis performed"
    
    # 4. Metadata integrity score
    metadata_score, metadata_explanation = calculate_metadata_integrity_score(evidence_list)
    
    # Calculate weighted final score
    weights = {
        "location": settings.LOCATION_WEIGHT,
        "time": settings.TIME_WEIGHT,
        "evidence_type": settings.EVIDENCE_TYPE_WEIGHT,
        "visual": settings.VISUAL_RELEVANCE_WEIGHT,
        "metadata": settings.METADATA_INTEGRITY_WEIGHT
    }
    
    final_score = (
        location_score * weights["location"] +
        time_score * weights["time"] +
        evidence_type_score * weights["evidence_type"] +
        avg_visual_score * weights["visual"] +
        metadata_score * weights["metadata"]
    )
    
    # Clamp to 0-100
    final_score = max(0.0, min(100.0, final_score))
    
    # Determine status based on score ranges
    if final_score >= 75:
        status = ClaimStatus.APPROVED
        status_explanation = "High confidence - claim appears legitimate"
    elif final_score >= 50:
        status = ClaimStatus.REVIEW_REQUIRED
        status_explanation = "Medium confidence - manual review recommended"
    elif final_score >= 25:
        status = ClaimStatus.REVIEW_REQUIRED
        status_explanation = "Low confidence - thorough review required"
    else:
        status = ClaimStatus.REJECTED
        status_explanation = "Very low confidence - likely invalid claim"
    
    # Build final explanation
    final_explanation = f"{status_explanation}. Score breakdown: "
    final_explanation += f"Location ({location_score:.1f}), "
    final_explanation += f"Time ({time_score:.1f}), "
    final_explanation += f"Evidence Type ({evidence_type_score:.1f}), "
    final_explanation += f"Visual ({avg_visual_score:.1f}), "
    final_explanation += f"Metadata ({metadata_score:.1f})"
    
    # Create scoring factors
    factors = ScoringFactors(
        location_score=location_score,
        location_explanation=location_explanation,
        time_score=time_score,
        time_explanation=time_explanation,
        evidence_type_score=evidence_type_score,
        evidence_type_explanation=evidence_type_explanation,
        visual_relevance_score=avg_visual_score,
        visual_relevance_explanation=visual_explanation,
        metadata_integrity_score=metadata_score,
        metadata_integrity_explanation=metadata_explanation
    )
    
    return ClaimScore(
        confidence_score=final_score,
        status=status,
        factors=factors,
        final_explanation=final_explanation
    )